### ISSUE TITLE
Improve the robustness of astroquery requests

### ISSUE THREAD

**Issue by andamian** [astropy/astroquery#1830]
I'm proposing a few improvements to deal with networking errors in CADC and ALMA and  possibly `astroquery` in general.

- Re-try mechanism for transient errors. Some servers might respond with an Http 503 error when overloaded or temporarily offline. The library should retry when encounters these errors. 503 even has a mechanism for this where the server could send a `Retry-After` field in the response. Example of such errors:
            `requests.codes.unavailable,
            requests.codes.service_unavailable,
            requests.codes.gateway_timeout,
            requests.codes.request_timeout,
            requests.codes.timeout,
            requests.codes.precondition_failed,
            requests.codes.precondition,
            requests.codes.payment_required,
            requests.codes.payment`
- Ensure that the transfer is complete when downloading data and skip downloads that are not necessary (source and destination are the same). It could use both the size and the checksum to verify that. Right now the size of the downloaded file does not seem to be checked against the `Content-Length` of the file.
- Check service is up. IVOA services have a discoverable end point to checking the status of a service. 

Before I dive too deep into the implementation, please comment on these proposed improvements. Maybe they are in but I'm missing them.

Thanks

**Comment by keflavich** [comment]
We do have the middle one:
https://github.com/astropy/astroquery/blob/c28ff416fbcbc9b12c5e9a082b48bfcc6186356a/astroquery/query.py#L307

**Comment by keflavich** [comment]
We have a really simplistic retry implemented in NRAO, but I think we could use that in `astroquery.query` too.

The "is it up" checks would be *great*, I didn't know any of those exist.

**Comment by bsipocz** [comment]
@andamian  - these all sounds great to me. I agree with @keflavich that the second in the list should be covered, but improvements to the caching is a regularly raised item.

**Comment by RilyndC** [comment]
Checking main, BaseQuery.__init__ still constructs a plain requests.Session() with no HTTPAdapter mounted, so there's no retry at any layer. Transient connection resets surface to users as a raw urllib3 traceback with nothing indicating the failure is transient or worth retrying.

Would a PR mounting an HTTPAdapter with a urllib3 Retry on the session be welcome? Rough shape: retry on connection errors and 5xx only, never on 4xx, with exponential backoff and a low total so a real outage still fails fast rather than hanging.

Happy to take a steer on whether this belongs session-wide in BaseQuery or opt-in per module.

**Comment by keflavich** [comment]
> Would a PR mounting an HTTPAdapter with a urllib3 Retry on the session be welcome? 

No, this is not a high-demand feature.  I'd rather have @andamian lead this.

**Comment by RilyndC** [comment]
Understood, thanks for the quick response.
